test/protect.jl

```diff
@@ -0,0 +1,43 @@
+@testset "Protection" begin
+    d = (a=rand(10), b=[1:10;], c=rand(Int, 10))
+
+    @testset "protect" begin
+        # protect blocks interpreting +, &, and * as formula specials
+        f = @formula(0 ~ protect(a+b) + a + b)
+        ff = apply_schema(f, schema(d))
+        @test ff.rhs.terms[1] isa FunctionTerm{typeof(+)}
+        @test modelcols(ff.rhs, d) == [d.a .+ d.b d.a d.b]
+
+        # protect is applied by default inside FunctionTerm args
+        f = @formula(0 ~ 1 - protect(a+b))
+        ff = apply_schema(f, schema(d))
+        ff2 = apply_schema(@formula(0 ~ 1 - (a+b)), schema(d))
+        @test modelcols(ff.rhs, d) == modelcols(ff2.rhs, d) ≈ 1 .- (d.a .+ d.b)
+    end
+
+    @testset "unprotect" begin
+        # unprotect reverts to treating calls to +, &, and * as term union,
+        # interaction, and combined
+
+        f = @formula(0 ~ 1 - unprotect(a&b))
+        ff = apply_schema(f, schema(d))
+        @test modelcols(ff.rhs, d) ≈ 1 .- d.a .* d.b
+
+        # ideally you'd also be able to do these but it's hard to make it work...
+        # this fails because - doesn't auto-broadcast, and the broadcasting that
+        # happens during FunctionTerm evaluation gets used up by the (a,b) tuple.
+        ff = apply_schema(@formula(0 ~ 1 - unprotect(a+b)), schema(d))
+        @test_broken modelcols(ff.rhs, d) == 1 .- [d.a d.b]
+
+        # and even if we define a broadcasting version, still fails because of
+        # packing the result into a right-sized container.  end up with a tuple
+        # of arrays which needs to be collected into a matrix
+        my_sub = (x,y) -> x .- y
+        ff = apply_schema(@formula(0 ~ my_sub(1, unprotect(a+b))), schema(d))
+        @test_broken modelcols(ff.rhs, d) == 1 .- [d.a d.b]
+        
+        # both of these could be fixed by always returning a matrix when you call
+        # modelcols on a tuple of terms but that would break other things
```

Comment:
Do you think packages may depend on this?

Reply:
yes, at least MixedModels.jl does!  it relies on the distinction between ranef terms being "non matrix terms" and fixef terms being "matrix terms" to create a tuple of `(allfixefterms_matrix, ranef1_matrix, ranef2_matrix, ...)`

Reply:
MAYBE we coudl work around that by using some kind of "lazy model matrix" system where you don't actually collect everything into a `Matrix`

Reply:
I'm a bit lost... What's the relationship between this and `unprotect`? Is the concern only that it's breaking, or also that the API isn't completely satisfying?

Reply:
The main concern is that it's breaking; the most salient case off the top of my head is that MixedModels uses `extract_matrix_terms` to separate out the fixed effect terms from the random effects terms, and then calls `modelcols` on teh tuple of terms that results.

Reply:
Sorry, don't have a clear view of all this. Anyway I guess the two main questions are: does this PR break MixedModels? does it allow fixing this later without breaking the API?